File: kernels/Julia/Compositron/src/PALS/PALSpectrum.jl
using Format

using ..Utils
using ..DBS
using ..Constants

export PALSpectrum, fit!, fit_report


mutable struct PALSpectrum
    spectrum::Vector{<:Unsigned}
    detpair::TimingDetectorPair
    counts::UInt64
    peak_params::Dict{String, SimpleFitParam}
    fit_result::Union{Nothing, FitResult}

    function PALSpectrum(
        spectrum::AbstractVector{<:Integer}, detpair::TimingDetectorPair
    )
        new(
            min_spectrum(spectrum),
            detpair,
            sum(spectrum),
            Dict{String, SimpleFitParam}(),
            nothing,
        )
    end
end


function _get_peak_center!(p::PALSpectrum)::Float64
    max_idx = argmax(p.spectrum)

    peak_center_window = 15

    a = max(max_idx - peak_center_window, 0)
    b = min(max_idx + peak_center_window, length(p.spectrum))

    y = Float64.(p.spectrum[a:b])
    x = Float64.(a:b)

    params = fit_gauss(x, y, [maximum(y), peak_center_window, 100])

    peak_height = params["amp_1"]
    peak_center = params["x0_1"]
    peak_width = params["sig_1"]

    p.peak_params["center_idx"] = peak_center
    p.peak_params["height"] = peak_height
    p.peak_params["fwhm"] = SimpleFitParam(
        peak_width.val * FWHM_OVER_SIGMA, peak_width.err * FWHM_OVER_SIGMA,
    )

    peak_center.val + a
end


function fit!(
    p::PALSpectrum,
    model::LifetimeModel,
    fit_start_idx::Integer,
    fit_end_idx::Integer,
)
    peak_center = if haskey(p.peak_params, "center_idx")
        p.peak_params["center_idx"]
    else
        _get_peak_center!(p)
    end

    y = Float64.(p.spectrum[fit_start_idx:fit_end_idx])
    x = fit_start_idx:fit_end_idx .|> i -> from_index(p.detpair.tcal, i)

    p.fit_result = fit_lifetime_spectrum(
        x, y, model, p.counts, from_index(p.detpair.tcal, peak_center)
    )
end


function _print_params(params::Vector{FitParam}, names::Vector{String})::String
    text = ""
    n_rows = length(names)

    values = [
        begin
            if isnan(param.val)
                "NaN"
            elseif startswith(name, "int")
                format("{:.2f}", 100 * param.val)
            elseif abs(param.val) >= 1e4
                format("{:.3e}", param.val)
            else
                format("{:.2f}", param.val)
            end
        end
        for (name, param) in zip(names, params)
    ]

    abs_errors = [
        begin
            if isnan(param.err)
                "NaN"
            elseif startswith(name, "int")
                format("{:.1f}", 100 * param.err)
            elseif abs(param.err) >= 1e4
                format("{:.3e}", param.err)
            else
                format("{:.2f}", param.err)
            end
        end
        for (name, param) in zip(names, params)
    ]

    rel_errors = [
        begin
            x = abs(param.err / param.val * 100)
            if isnan(x)
                "NaN %"
            else
                format("{:.2f} %", x)
            end
        end
        for param in params
    ]

    mins = [
        begin
            if isnan(param.min)
                "NaN"
            elseif isinf(param.min)
                if param.min > 0
                    "Inf"
                else
                    "-Inf"
                end
            elseif startswith(name, "int")
                format("{:.1f}", 100 * param.min)
            else
                format("{:.1f}", param.min)
            end
        end
        for (name, param) in zip(names, params)
    ]

    maxs = [
        begin
            if isnan(param.max)
                "NaN"
            elseif isinf(param.min)
                if param.min > 0
                    "Inf"
                else
                    "-Inf"
                end
            elseif startswith(name, "int")
                format("{:.1f}", 100 * param.max)
            else
                format("{:.1f}", param.max)
            end
        end
        for (name, param) in zip(names, params)
    ]

    fixed = [
        begin
            if param.vary
                ""
            else
                "Fixed"
            end
        end
        for param in params
    ]

    at_bnd = [
        begin
            if abs(param.val - param.min) < 1e-3 || abs(param.val - param.max) < 1e-3
                "At Boundary"
            else
                ""
            end
        end
        for param in params
    ]

    columns = [
        names, values, abs_errors, rel_errors, mins, maxs, fixed, at_bnd
    ]

    col_widths = [maximum(length(row) for row in col) for col in columns]

    headers = [
        "Parameter", "Value", "Stderr (abs)", "Stderr (rel)", "Min", "Max",
        "Fixed", ""
    ]

    head_widths = [length(head) for head in headers]

    widths = [max(c, h) for (c, h) in zip(col_widths, head_widths)]

    formats = ["<", ">", ">", ">", ">", ">", "<", "<"]

    for (i, head) in enumerate(headers)
        text *= format(format(" {{:^{}}} ", widths[i]), head)
    end

    text *= "\n"

    for wid in widths
        text *= format(format(" {{:-^{}}} ", wid), "")
    end

    text *= "\n"

    for row in 1:n_rows
        for (col, column) in enumerate(columns)
            if formats[col] == "<"
                text *= format(format(" {{:<{}}} ", widths[col]), column[row])
            else
                text *= format(format(" {{:>{}}} ", widths[col]), column[row])
            end
        end
        text *= "\n"
    end

    text
end


function fit_report(p::PALSpectrum)
    if isnothing(p.fit_result)
        throw(ErrorException("No fit has been performed yet"))
    end

    text = ""

    result = something(p.fit_result)
    model = result.model

    n_l = length(model.lifetime_components)
    n_r = length(model.resolution_components)

    thick_hline = format("{:#^100}\n", "")
    thin_hline = format("{:-^100}\n", "")

    text *= thick_hline
    text *= format("{:^100}\n", "Fit report")
    text *= thick_hline
    text *= "\n"
    text *= format("Statistics: {}\n", p.counts)
    text *= format("Number of Data Points: {}\n", result.n_dpoints)
    text *= "\n"
    text *= format("Lifetime Components:   {}\n", n_l)
    text *= format("Resolution Components: {}\n", n_r)
    if model.background_component.vary
        text *= "Background Components: 1\n"
    else
        text *= "Background Components: 0\n"
    end
    text *= "\n"
    text *= "Fit Method: Trust-Region\n"
    text *= format("Fit Status: {}\n", result.fit_status)
    text *= format("Number of Function Evaluations: {}\n", result.n_eval)
    if isnothing(result.red_chi_2)
        text *= "Reduced Chi squared: NaN\n"
    else
        text *= format("Reduced Chi squared: {}\n", result.red_chi_2)
    end
    text *= "\n"
    text *= thin_hline
    text *= format("{:^100}\n", "Parameters")
    text *= thin_hline
    text *= "\n"
    text *= format("{:=^100}\n", " General Parameters ")
    text *= "\n"

    text *= _print_params(
        [model.scale_component, model.shift_component, model.background_component],
        ["N", "t0", "background"]
    )

    text *= "\n"
    text *= format("{:=^100}\n", " Lifetime Components ")
    text *= "\n"

    lt_norm = sum(comp.intensity.val for comp in model.lifetime_components)

    if abs(lt_norm  - 1) > 1e-3
        text *= "!!! intensities don't add up to 100% !!!\n\n"
    end

    params = Vector{FitParam}()
    names = Vector{String}()

    for (i, comp) in enumerate(model.lifetime_components)
        push!(params, comp.lifetime)
        push!(names, "lifetime_$i")
    end
    for (i, comp) in enumerate(model.lifetime_components)
        push!(params, comp.intensity)
        push!(names, "intensity_$i")
    end

    text *= _print_params(params, names)

    text *= "\n"
    text *= format("{:=^100}\n", " Resolution Components ")
    text *= "\n"

    res_norm = sum(comp.intensity.val for comp in model.resolution_components)

    if abs(res_norm  - 1) > 1e-3
        text *= "!!! intensities don't add up to 100% !!!\n\n"
    end

    params = Vector{FitParam}()
    names = Vector{String}()

    for (i, comp) in enumerate(model.resolution_components)
        push!(params, comp.fwhm)
        push!(names, "fwhm_$i")
    end
    for (i, comp) in enumerate(model.resolution_components)
        push!(params, comp.intensity)
        push!(names, "intensity_$i")
    end
    for (i, comp) in enumerate(model.resolution_components)
        push!(params, comp.t0)
        push!(names, "t0_$i")
    end

    text *= _print_params(params, names)
    text *= "\n"
    text *= thin_hline

    text
end

File: kernels/Rust/compositron/src/pals/palspectrum.rs
use std::collections::HashMap;

use thiserror::Error;

use crate::constants::FWHM_OVER_SIGMA;
use crate::core::utils::{spectrum_match, TimingDetectorPair, Spectrum};
use crate::core::fitting::{FitParam, LMFitError, SimpleFitParam, VarproFitError};
use crate::dbs::fitting::fit_gauss;
use crate::pals::fitting::{fit_lifetime_spectrum, FitResult};
use crate::pals::model::*;


#[derive(Debug, Error)]
pub enum AnalysisError {
    #[error("Fit failed")]
    LMFitError {
        #[from]
        source: LMFitError
    },

    #[error("Fit failed")]
    VarproFitError {
        #[from]
        source: VarproFitError
    },
}


#[derive(Debug, Error)]
pub enum ReportError {
    #[error("No fit has been performed yet")]
    NoFitPerformed,

    #[error("I/O Error")]
    IOError {
        #[from]
        source: std::fmt::Error,
    }
}


pub struct PALSpectrum {
    pub spectrum: Spectrum,
    pub detpair: TimingDetectorPair,
    pub counts: u64,
    pub peak_params: HashMap<&'static str, SimpleFitParam>,
    pub fit_result: Option<FitResult>,
}


impl PALSpectrum {
    pub fn new(
        spectrum: Spectrum,
        detpair: TimingDetectorPair,
    ) -> Self {
        let counts = spectrum_match!(
            &spectrum, arr => arr.iter().map(|&x| x as u64).sum::<u64>()
        );

        PALSpectrum {
            spectrum,
            detpair,
            counts,
            peak_params: HashMap::new(),
            fit_result: None,
        }
    }


    fn get_peak_center(&mut self) -> Result<f64, AnalysisError> {
        let (argmax, _) = spectrum_match!(
            &self.spectrum, arr => arr
                .iter()
                .enumerate()
                .map(|(i, &y)| (i, y as u64))
                .max_by_key(|&(_, y)| y)
                .unwrap_or((0, 0))
        );

        let peak_center_window = 15;

        let roi = argmax - peak_center_window .. argmax + peak_center_window;

        let y = spectrum_match!(
            &self.spectrum, arr => arr.rows(roi.start, roi.len())
                .iter()
                .map(|&i| i as f64)
                .collect::<Vec<f64>>()
        );

        let x = roi.map(|i| i as f64).collect::<Vec<f64>>();

        let mut peak_params = fit_gauss(&x, &y, vec![argmax as f64, 100.])?;

        let peak_height = peak_params.remove("amp_1").unwrap();
        let peak_center = peak_params.remove("x0_1").unwrap();
        let peak_width = peak_params.remove("sig_1").unwrap();

        let x0 = peak_center.val;

        self.peak_params.insert("center_idx", peak_center);
        self.peak_params.insert("height", peak_height);
        self.peak_params.insert("fwhm", SimpleFitParam {
            val: peak_width.val * FWHM_OVER_SIGMA,
            err: peak_width.err * FWHM_OVER_SIGMA
        });

        Ok(x0)
    }


    pub fn fit(
        &mut self,
        model: LifetimeModel,
        fit_start_idx: usize,
        fit_end_idx: usize,
    ) -> Result<(), AnalysisError> {
        let peak_center = self.peak_params
            .get("center_idx")
            .map(|p| p.val)
            .unwrap_or(self.get_peak_center()?);

        let roi = fit_start_idx..fit_end_idx;

        let y = spectrum_match!(
            &self.spectrum, arr => arr.rows(roi.start, roi.len())
                .iter()
                .map(|&i| i as f64)
                .collect::<Vec<f64>>()
        );

        let tcal = self.detpair.corrected_tcal.unwrap_or(self.detpair.tcal);

        let x = roi.map(|i| tcal.from_index(i)).collect::<Vec<f64>>();

        let fit_result = fit_lifetime_spectrum(
            &x,
            &y,
            model,
            self.counts as f64,
            tcal.from_index_f64(peak_center),
        )?;

        self.fit_result = Some(fit_result);

        Ok(())
    }


    fn print_params(params: &[&FitParam], names: Vec<String>) -> String {
        let mut text = String::new();
        let n_rows = names.len();

        let values = params
            .iter()
            .zip(&names)
            .map(|(param, name)| {
                if name.starts_with("int") {
                    format!("{:.2}", 100. * param.val)
                } else if param.val.abs() >= 1e4 {
                    format!("{:.3e}", param.val)
                } else {
                    format!("{:.2}", param.val)
                }
            })
            .collect::<Vec<String>>();

        let abs_errors = params
            .iter()
            .zip(&names)
            .map(|(param, name)| {
                if name.starts_with("int") {
                    format!("{:.1}", 100. * param.err)
                } else if param.err.abs() >= 1e4 {
                    format!("{:.3e}", param.err)
                } else {
                    format!("{:.2}", param.err)
                }
            })
            .collect::<Vec<String>>();

        let rel_errors = params
            .iter()
            .map(|param| format!("{:.2} %", (param.err / param.val * 100.).abs()))
            .collect::<Vec<String>>();

        let mins = params
            .iter()
            .zip(&names)
            .map(|(param, name)| {
                if name.starts_with("int") {
                    format!("{:.1}", 100. * param.min)
                } else {
                    format!("{:.1}", param.min)
                }
            })
            .collect::<Vec<String>>();


        let maxs = params
            .iter()
            .zip(&names)
            .map(|(param, name)| {
                if name.starts_with("int") {
                    format!("{:.1}", 100. * param.max)
                } else {
                    format!("{:.1}", param.max)
                }
            })
            .collect::<Vec<String>>();

        let fixed = params
            .iter()
            .map(|param| {
                if param.vary {
                    "".to_string()
                } else {
                    "Fixed".to_string()
                }
            })
            .collect::<Vec<String>>();

        let at_bnd = params
            .iter()
            .map(|param| {
                if (param.val - param.min).abs() < 1e-3 || (param.val - param.max).abs() < 1e-3 {
                    "At Boundary".to_string()
                } else {
                    "".to_string()
                }
            })
            .collect::<Vec<String>>();

        let columns = [
            names, values, abs_errors, rel_errors, mins, maxs, fixed, at_bnd
        ];

        let col_widths = columns
            .iter()
            .map(|arr| arr.iter().map(|x| x.len()).max().unwrap_or(0))
            .collect::<Vec<usize>>();

        let headers = [
            "Parameter", "Value", "Stderr (abs)", "Stderr (rel)", "Min", "Max",
            "Fixed", ""
        ];

        let head_widths = headers
            .iter()
            .map(|head| head.len())
            .collect::<Vec<usize>>();

        let widths = col_widths
            .iter()
            .zip(&head_widths)
            .map(|(c, h)| *c.max(h))
            .collect::<Vec<usize>>();

        let formats = ["<", ">", ">", ">", ">", ">", "<", "<"];

        for (i, head) in headers.iter().enumerate() {
            text.push_str(&format!(" {:^1$} ", head, widths[i]));
        }

        text.push('\n');

        for wid in &widths {
            text.push_str(&format!(" {:-^1$} ", "", wid));
        }

        text.push('\n');

        for row in 0..n_rows {
            for (col, column) in columns.iter().enumerate() {
                if formats[col] == "<" {
                    text.push_str(&format!(" {:<1$} ", column[row], widths[col]));
                } else {
                    text.push_str(&format!(" {:>1$} ", column[row], widths[col]));
                }
            }
            text.push('\n');
        }

        text
    }


    pub fn fit_report(
        &self,
    ) -> Result<String, ReportError> {
        let mut text = String::new();

        let Some(result) = &self.fit_result else {
            return Err(ReportError::NoFitPerformed);
        };

        let model = &result.model;

        let n_l = model.lifetime_components.len();
        let n_r = model.resolution_components.len();

        let thick_hline = format!("{:#^100}\n", "");
        let thin_hline = format!("{:-^100}\n", "");

        text.push_str(&thick_hline);
        text.push_str(&format!("{:^100}\n", "Fit report"));
        text.push_str(&thick_hline);
        text.push('\n');
        text.push_str(&format!("Statistics: {}\n", self.counts));
        text.push_str(&format!("Number of Data Points: {}\n", result.n_dpoints));
        text.push('\n');
        text.push_str(&format!("Lifetime Components:   {}\n", n_l));
        text.push_str(&format!("Resolution Components: {}\n", n_r));
        if model.background_component.default_initialized {
            text.push_str("Background Components: 0\n");
        } else {
            text.push_str("Background Components: 1\n");
        }
        text.push('\n');
        text.push_str("Fit Method: Levenberg-Marquardt\n");
        text.push_str("Fit Status: Success\n");
        text.push_str(&format!("Reason for Termination: {}\n", result.fit_status));
        text.push_str(&format!("Number of Function Evaluations: {}\n", result.n_eval));
        if let Some(red_chi_2) = result.red_chi_2 {
            text.push_str(&format!("Reduced Chi squared: {}\n", red_chi_2));
        } else {
            text.push_str("Reduced Chi squared: NaN\n");
        }
        text.push('\n');
        text.push_str(&thin_hline);
        text.push_str(&format!("{:^100}\n", "Parameters"));
        text.push_str(&thin_hline);
        text.push('\n');
        text.push_str(&format!("{:=^100}\n", " General Parameters "));
        text.push('\n');

        text.push_str(&PALSpectrum::print_params(
            &[&model.scale_component, &model.shift_component, &model.background_component],
            vec!["N".into(), "t0".into(), "background".into()],
        ));

        text.push('\n');
        text.push_str(&format!("{:=^100}\n", " Lifetime Parameters "));
        text.push('\n');

        let lt_norm = model.lifetime_components.iter().map(|comp| comp.intensity.val).sum::<f64>();

        if (lt_norm - 1.).abs() > 1e-3 {
            text.push_str("!!! intensities don't add up to 100% !!!\n\n");
        }

        let mut params = Vec::new();
        let mut names = Vec::new();

        for i in 0..n_l {
            params.push(&model.lifetime_components[i].lifetime);
            names.push(format!("lifetime_{}", i+1));
        }
        for i in 0..n_l {
            params.push(&model.lifetime_components[i].intensity);
            names.push(format!("intensity_{}", i+1));
        }

        text.push_str(&PALSpectrum::print_params(&params, names));

        text.push('\n');
        text.push_str(&format!("{:=^100}\n", " Resolution Parameters "));
        text.push('\n');

        let res_norm = model.resolution_components.iter().map(|comp| comp.intensity.val).sum::<f64>();

        if (res_norm - 1.).abs() > 1e-3 {
            text.push_str("!!! Intensities don't add up to 100% !!!\n");
        }

        let mut params = Vec::new();
        let mut names = Vec::new();

        for i in 0..n_r {
            params.push(&model.resolution_components[i].fwhm);
            names.push(format!("fwhm_{}", i+1));
        }
        for i in 0..n_r {
            params.push(&model.resolution_components[i].intensity);
            names.push(format!("intensity_{}", i+1));
        }
        for i in 0..n_r {
            params.push(&model.resolution_components[i].t0);
            names.push(format!("t0_{}", i+1));
        }

        text.push_str(&PALSpectrum::print_params(&params, names));
        text.push('\n');
        text.push_str(&thin_hline);

        Ok(text)
    }
}
